src/main/java/net/minecraftforge/registries/NamespacedHolderHelper.java

```diff
@@ -88,15 +90,31 @@ Optional<Holder<T>> getHolder(T value)
         return Optional.ofNullable(this.holders.get(value));
     }
 
-    Holder<T> getOrCreateHolder(ResourceKey<T> key)
+    DataResult<Holder<T>> getOrCreateHolder(ResourceKey<T> key)
     {
-        return this.holdersByName.computeIfAbsent(key.location(), k ->
+        Holder.Reference<T> ref = this.holdersByName.get(key.location());
+        if (ref == null)
         {
             if (this.holderLookup != null)
-                throw new IllegalStateException("This registry can't create new holders without value");
+                return DataResult.error("This registry can't create new holders without value (requested key: " + key + ")");
+
             if (this.frozen)
-                throw new IllegalStateException("Registry is already frozen (trying to add key " + k + ")");
-            return Holder.Reference.createStandAlone(this.self, key);
+                return DataResult.error("Registry is already frozen (trying to add key " + key + ")");
+
+            ref = Holder.Reference.createStandAlone(this.self, key);
+            this.holdersByName.put(key.location(), ref);
+        }
+        return DataResult.success(ref);
+    }
+
+    Holder<T> getOrCreateHolderOrThrow(ResourceKey<T> key)
+    {
+        return LamdbaExceptionUtils.uncheck(() ->
+        {
+            DataResult<Holder<T>> dataResult = getOrCreateHolder(key);
+            if (dataResult.error().isPresent())
+                throw new IllegalStateException(dataResult.error().get().message());
+            return dataResult.result().get();
```

Comment:
You may have misunderstood what I meant by using `LamdbaExceptionUtils`. In this form,  `LamdbaExceptionUtils#uncheck` effectively does nothing as `IllegalStateException` is already an unchecked exception. What I meant to suggest was to call `DataResult#getOrThrow` like so:
```java
return getCreateHolder(key).getOrThrow(false, LamdbaExceptionUtils.rethrowConsumer((String message) -> {
    throw new IllegalStateException(message);
}));
```

Reply:
Is it even necessary in this form?

Reply:
Actually, that's a fair point which I hadn't considered 👀 -- we don't even need `LamdbaExceptionUtils` because `IllegalStateException` is an unchecked exception.

Reply:
Yes, unfortunately, misunderstood its use. Figured we only wanted the illegal state, but I suppose it doesn't impact anything if we throw both. I'll push the revision now.